libraries/chain/transaction_metadata.cpp

```diff
@@ -4,35 +4,32 @@
 
 namespace eosio { namespace chain {
 
-
-const flat_set<public_key_type>& transaction_metadata::recover_keys( const chain_id_type& chain_id ) {
+recovery_keys_type transaction_metadata::recover_keys( const chain_id_type& chain_id ) {
    // Unlikely for more than one chain_id to be used in one nodeos instance
-   if( !signing_keys || signing_keys->first != chain_id ) {
-      if( signing_keys_future.valid() ) {
-         std::tuple<chain_id_type, fc::microseconds, flat_set<public_key_type>> sig_keys = signing_keys_future.get();
-         if( std::get<0>( sig_keys ) == chain_id ) {
-            sig_cpu_usage = std::get<1>( sig_keys );
-            signing_keys.emplace( std::get<0>( sig_keys ), std::move( std::get<2>( sig_keys )));
-            return signing_keys->second;
-         }
+   if( signing_keys_future.valid() ) {
+      const std::tuple<chain_id_type, fc::microseconds, flat_set<public_key_type>>& sig_keys = signing_keys_future.get();
+      if( std::get<0>( sig_keys ) == chain_id ) {
+         return std::make_pair( std::get<1>( sig_keys ), std::cref( std::get<2>( sig_keys ) ) );
       }
-      flat_set<public_key_type> recovered_pub_keys;
-      sig_cpu_usage = packed_trx->get_signed_transaction().get_signature_keys( chain_id, fc::time_point::maximum(), recovered_pub_keys );
-      signing_keys.emplace( chain_id, std::move( recovered_pub_keys ));
+      EOS_ASSERT( false, chain_id_type_exception, "chain id ${cid} does not match start_recover_keys ${sid}",
```

Comment:
Now that I think about it more, I strongly prefer to maintain the old behavior of always returning a valid result rather than throwing an exception. That way even if it was used incorrectly somehow in the code (i.e. not calling `start_recover_keys` before), we know it wouldn't throw an exception when before it didn't (and thus be a consensus break).

So, if the future is not valid or it doesn't match the provided `chain_id`, it should replace the future with a new one for the requested `chain_id` and then immediately synchronously recover the keys with no deadline and set the results into the future, before then grabbing it out from the future to return.